File: examples/flutter/duis06_prop_drilling_bad.dart
import 'package:flutter/material.dart';

/// DUIS-06 – Prop Drilling / Poor Data Flow
/// - themeColor is passed through intermediate widgets that don't use it.
class PropDrillingRoot extends StatelessWidget {
  const PropDrillingRoot({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: LevelOne(themeColor: Colors.blue),
    );
  }
}

class LevelOne extends StatelessWidget {
  final Color themeColor;
  const LevelOne({super.key, required this.themeColor});

  @override
  Widget build(BuildContext context) {
    return LevelTwo(themeColor: themeColor); // just forwards
  }
}

class LevelTwo extends StatelessWidget {
  final Color themeColor;
  const LevelTwo({super.key, required this.themeColor});

  @override
  Widget build(BuildContext context) {
    return LevelThree(themeColor: themeColor); // just forwards
  }
}

class LevelThree extends StatelessWidget {
  final Color themeColor;
  const LevelThree({super.key, required this.themeColor});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Text(
        "Deep child",
        style: TextStyle(color: themeColor),
      ),
    );
  }
}

File: examples/compose/duis06_prop_drilling_bad.kt
package examples.compose

import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color

/**
 * DUIS-06 – PROP DRILLING
 *
 * A color (themeColor) is passed through intermediate composables that don't use it,
 * only to reach a deep child. Similar to React's prop drilling.
 */

@Composable
fun PropDrillingRoot() {
    Scaffold(
        topBar = { TopAppBar(title = { Text("Prop Drilling Example") }) }
    ) { padding ->
        LevelOne(themeColor = Color.Blue, modifier = Modifier.padding(padding))
    }
}

@Composable
fun LevelOne(themeColor: Color, modifier: Modifier = Modifier) {
    // Does not use themeColor, just forwards
    LevelTwo(themeColor = themeColor, modifier = modifier)
}

@Composable
fun LevelTwo(themeColor: Color, modifier: Modifier = Modifier) {
    // Still just forwards
    LevelThree(themeColor = themeColor, modifier = modifier)
}

@Composable
fun LevelThree(themeColor: Color, modifier: Modifier = Modifier) {
    Text(
        text = "Deep child using themeColor",
        color = themeColor,
        modifier = modifier
    )
}
